elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/metadata/FormulaValidator.java

```diff
@@ -5,95 +5,128 @@
 
     private static String getColumnId(Queryable parent, ColumnProjection column) {
-        return parent.getName() + "." + column.getName();
+        return parent.getName() + PERIOD + column.getName();
     }
 
-    public FormulaValidator(MetaDataStore metaDataStore) {
-        super(metaDataStore);
-    }
+    public FormulaValidator(TableContext tableCtx) {
+        this.tblCtx = tableCtx;
 
-    @Override
-    protected Void visitFormulaMetric(Queryable parent, MetricProjection metric) {
-        return visitFormulaColumn(parent, metric);
-    }
+        this.handlebars = new Handlebars()
+                        .with(EscapingStrategy.NOOP)
+                        .registerHelper("sql", new Helper<Object>() {
+
+                            @Override
+                            public Object apply(final Object context, final Options options) throws IOException {
+                                String from = options.hash("from");
+                                String column = options.hash("column");
+                                int argsIndex = column.indexOf('[');
+
+                                // Remove args from column
+                                column = argsIndex == -1 ? column : column.substring(0, argsIndex);
+                                // Prefix column with join table name
+                                column = nullOrEmpty(from) ? column : from + PERIOD + column;
 
-    @Override
-    protected Void visitFormulaDimension(Queryable parent, ColumnProjection dimension) {
-        return visitFormulaColumn(parent, dimension);
+                                return column;
+                            }
+                        });
     }
 
-    @Override
-    protected Void visitFieldDimension(Queryable parent, ColumnProjection dimension) {
-        return null;
+    public void validateColumn(Queryable source, ColumnProjection column) {
+        checkCycle(tblCtx, column.getName(), getColumnId(source, column), new HashSet<>());
     }
 
-    /**
-     * For a FORMULA column, we just need to check all source columns in the formula expression.
-     *
-     * @param column a column defined with {@link MetricFormula} or {@link DimensionFormula}
-     * @return null
-     */
-    private Void visitFormulaColumn(Queryable source, ColumnProjection column) {
-        if (visited.contains(getColumnId(source, column))) {
-            throw new IllegalArgumentException(referenceLoopMessage(visited, source, column));
+    private void checkCycle(TableContext tblCtx, String columnName, String columnId, Set<String> alreadyVisited) {
+
+        Set<String> visited = new HashSet<>(alreadyVisited);
+
+        // Physical References starts with $
+        if (columnName.lastIndexOf('$') == 0) {
+            return;
         }
 
-        Type<?> tableClass = dictionary.getEntityClass(source.getName(), source.getVersion());
+        if (!visited.add(tblCtx.getAlias() + PERIOD + columnName)) {
```

Comment:
if I have a cycle A -> B -> A
won't my aliases be:
A,
A.B
A.B.A
And no cycle found?